z3/src/tactic.rs

```diff
@@ -0,0 +1,191 @@
+use std::ffi::{CStr, CString};
+use std::fmt;
+use std::result::Result;
+use std::str::Utf8Error;
+
+use z3_sys::*;
+use Context;
+use Goal;
+use Params;
+use Tactic;
+use Z3_MUTEX;
+
+#[derive(Clone, Debug)]
+pub struct ApplyResult<'ctx> {
+    ctx: &'ctx Context,
+    z3_apply_result: Z3_apply_result,
+}
+
+impl<'ctx> ApplyResult<'ctx> {
+    pub fn list_subgoals(&self) -> impl Iterator<Item = Goal> {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
```

Comment:
this guard should be the first line in the map closure body, otherwise it would be dropped after this method returns